src/shims/foreign_items.rs

```diff
@@ -111,25 +115,190 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     /// This function will handle `goto_block` if needed.
     fn emulate_foreign_item(
         &mut self,
-        def_id: DefId,
+        link_name: &str,
         args: &[OpTy<'tcx, Tag>],
         dest: Option<PlaceTy<'tcx, Tag>>,
         ret: Option<mir::BasicBlock>,
+        unwind: Option<mir::BasicBlock>
     ) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
-        let attrs = this.tcx.get_attrs(def_id);
-        let link_name = match attr::first_attr_value_str_by_name(&attrs, sym::link_name) {
-            Some(name) => name.as_str(),
-            None => this.tcx.item_name(def_id).as_str(),
-        };
-        // Strip linker suffixes (seen on 32-bit macOS).
-        let link_name = link_name.trim_end_matches("$UNIX2003");
-        let tcx = &{ this.tcx.tcx };
+        let tcx = &{this.tcx.tcx};
 
         // First: functions that diverge.
         match link_name {
-            "__rust_start_panic" | "panic_impl" => {
-                throw_unsup_format!("the evaluated program panicked");
+            "__rust_start_panic" => {
```

Comment:
So conceptually, this is a reimplemenation of [this method](https://github.com/rust-lang/rust/blob/9ebf47851a357faa4cd97f4b1dc7835f6376e639/src/libpanic_unwind/lib.rs#L101) inside Miri, right? We are doing all the same things; the only part that's different is that instead of calling `imp::panic` for the current platform we return `StartUnwind` during popping of a stack frame. The "reimplementation" of that method is spread across 2 files: the part before the call to `box_me_up` is here; the part after that call is in `pop_frame`, executed when `box_me_up` returns.

I think a high-level comment *somewhere* should explain this, probably at the beginning of this block here, before it goes into the weeds of how to exactly do that.

But, could we have a short discussion about ways to avoid this? Do we have the MIR for `__rust_start_panic` from the runtime available; is that something we can access in Miri at this point? If yes, I think we should try to redirect our call to that function instead of re-implementing it here. All we need to do is make sure that `imp::panic` does not call any of the crazy low-level stuff; we could submit our own Miri `imp` to `libpanic_unwind` that calls a magic intrinsic or so that triggers the final part of starting unwinding. Would that not be better?